arangod/VocBase/LogicalCollection.cpp

```diff
@@ -1065,3 +1120,33 @@ VPackSlice LogicalCollection::keyOptions() const {
   }
   return VPackSlice(_keyOptions->data());
 }
+
+void LogicalCollection::validators(VPackBuilder& b) const {
+  VPackArrayBuilder guard(&b);
+  auto vals = std::atomic_load_explicit(&_validators, std::memory_order_relaxed);
+  for(auto const& validator : *vals) {
+    validator->toVelocyPack(b);
+  }
+}
+
+Result LogicalCollection::validate(VPackSlice s, bool validate) const {
+  if (!validate) { return {}; }
+  auto vals = std::atomic_load_explicit(&_validators, std::memory_order_relaxed);
+  for(auto const& validator : *vals) {
+    if(!validator->validate(s, VPackSlice::noneSlice(), ValidatorOperation::INSERT)) {
+      return {TRI_ERROR_BAD_PARAMETER, "validation failed: " + validator->message() };
+    }
+  }
+  return {};
+}
+
+Result LogicalCollection::validate(VPackSlice new_, VPackSlice old_, bool validate) const {
+  if (!validate) { return {}; }
+  auto vals = std::atomic_load_explicit(&_validators, std::memory_order_relaxed);
+  for(auto const& validator : *vals) {
+    if(!validator->validate(new_, old_, ValidatorOperation::UPDATE)) {
```

Comment:
UPDATE is not always correct here. It could also be REPLACE. I guess the operation type should be passed into the method instead as a parameter. Btw. there already exists `TRI_voc_document_operation_e` with an enum for all CRUD operation types.